### DISCUSSION TITLE
Keycloak Cookies

### DISCUSSION THREAD

**Discussion by cprouveu** [dasniko/testcontainers-keycloak#113] [Q&A]
Hello, 
I'm trying to test our keycloak provider. (keycloak 22.0.1)
So I tried your test container.
I included my target/classes.

Now : for our custom api call we used the keycloak cookies session.
<img width="256" alt="Capture d’écran 2023-09-06 à 17 18 47" src="https://github.com/dasniko/testcontainers-keycloak/assets/21106757/07eabd6d-e0d7-40c1-85e0-203c641b033a">
But how found them in the keycloak container ?
I found : 
            Keycloak keycloakClient = keycloak.getKeycloakAdminClient();
            AccessTokenResponse accessTokenResponse = keycloakClient.tokenManager().getAccessToken();
But I don't see cookies.

Thanks for your help.

**Comment by cprouveu** [comment]
Update : 
I found for cookies : 
KEYCLOAK_XXXX
            Keycloak keycloakClient = keycloak.getKeycloakAdminClient();
            AccessTokenResponse accessTokenResponse = keycloakClient.tokenManager().getAccessToken();
            TokenVerifier<AccessToken> verifier = TokenVerifier.create(accessTokenResponse.getToken(), AccessToken.class);
            AccessToken accessToken = verifier.getToken();
            String sub = accessToken.getSubject();
            String sessionState = accessToken.getSessionState();

So I have now : 
            String cookie = "KEYCLOAK_IDENTITY=" + accessTokenResponse.getToken();
            cookie = cookie + "; KEYCLOAK_IDENTITY_LEGACY=" + accessTokenResponse.getToken();
            cookie = cookie + "; KEYCLOAK_SESSION_LEGACY=master/" + sub + "/" + sessionState ;
            cookie = cookie + "; KEYCLOAK_SESSION=master/" + sub + "/" + sessionState ;

But for AUTH_SESSION_ID and AUTH_SESSION_ID_LEGACY, I found in documentation the format : \<session-id>.\<owner-node-id>\
I have the session-id but how found 'owner-node-id' for your keycloak container?

**Reply by dasniko** [reply]
This project just provides the default Keycloak Container as a Testcontainer and provides additionally the default Keycloak Admin Java Client.
There's nothing special in this project on how to use these components. Please refer to the official documentation of your desired components on how to use them properly.

**Reply by cprouveu** [reply]
I didn't pass by cookies so I used bearer token and it's working.

**Comment by knighthedspi** [comment]
I think it is better to use the cookies response from the Keycloak than setting by yourself.
You can load the configuration at `keycloak.getAuthServerUrl() + "/realms/" + realmName + "/.well-known/openid-configuration"` then get the `authorization_endpoint` for the login url.
After that, access this url with query parameters as:

- login = true
- response_type = code
- client_id = your client id
- state = random UUID
- redirect_uri = your client redirect uri
- scope = openid

And use cookies get from the response for further requests
